File: src/main/java/arcana/utils/wrappers/ETWrapper.java
package arcana.utils.wrappers;

import arcana.Arcana;
import net.minecraft.block.Block;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityClassification;
import net.minecraft.entity.EntityType;
import net.minecraft.world.World;

public class ETWrapper<T extends Entity> {
    String regName;
    EntityType.Builder<T> builder;

    public ETWrapper(EntityType.IFactory<T> factory, EntityClassification cls, String registryName) {
        regName = registryName;
        builder = EntityType.Builder.of(factory, cls);
    }
    @SuppressWarnings("unchecked")
    public EntityType<T> get(){
        return (EntityType<T>) builder.build("").setRegistryName(Arcana.id, regName);
    }

    public ETWrapper<T> sized(float a, float b){
        builder.sized(a, b);
        return this;
    }

    public ETWrapper<T> noSummon() {
        builder.noSummon();
        return this;
    }

    public ETWrapper<T> noSave() {
        builder.noSave();
        return this;
    }

    public ETWrapper<T> fireImmune() {
        builder.fireImmune();
        return this;
    }

    public ETWrapper<T> immuneTo(Block... blocks) {
        builder.immuneTo(blocks);
        return this;
    }

    public ETWrapper<T> canSpawnFarFromPlayer() {
        builder.canSpawnFarFromPlayer();
        return this;
    }

    public ETWrapper<T> clientTrackingRange(int pRange) {
        builder.clientTrackingRange(pRange);
        return this;
    }

    public ETWrapper<T> updateInterval(int interval) {
        builder.updateInterval(interval);
        return this;
    }

    public ETWrapper<T> setUpdateInterval(int interval) {
        builder.setUpdateInterval(interval);
        return this;
    }

    public ETWrapper<T> setTrackingRange(int range) {
        builder.setTrackingRange(range);
        return this;
    }

    public ETWrapper<T> setShouldReceiveVelocityUpdates(boolean value) {
        builder.setShouldReceiveVelocityUpdates(value);
        return this;
    }

    public ETWrapper<T> setCustomClientFactory(java.util.function.BiFunction<net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity, World, T> customClientFactory) {
        builder.setCustomClientFactory(customClientFactory);
        return this;
    }
}

File: src/main/kotlin/arcana/utils/wrappers/ETWrapper.kt
package arcana.utils.wrappers

import arcana.Arcana
import net.minecraft.block.Block
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityClassification
import net.minecraft.entity.EntityType
import net.minecraft.world.World
import net.minecraftforge.fml.network.FMLPlayMessages.SpawnEntity
import java.util.function.BiFunction

class ETWrapper<T : Entity?>(cls: EntityClassification, var regName: String, factory: EntityType.IFactory<T>) {
    var builder: EntityType.Builder<T>

    init {
        builder = EntityType.Builder.of(factory, cls)
    }

    fun get(): EntityType<T> {
        return builder.build("").setRegistryName(Arcana.id, regName) as EntityType<T>
    }

    fun sized(a: Float, b: Float): ETWrapper<T> {
        builder.sized(a, b)
        return this
    }

    fun noSummon(): ETWrapper<T> {
        builder.noSummon()
        return this
    }

    fun noSave(): ETWrapper<T> {
        builder.noSave()
        return this
    }

    fun fireImmune(): ETWrapper<T> {
        builder.fireImmune()
        return this
    }

    fun immuneTo(vararg blocks: Block?): ETWrapper<T> {
        builder.immuneTo(*blocks)
        return this
    }

    fun canSpawnFarFromPlayer(): ETWrapper<T> {
        builder.canSpawnFarFromPlayer()
        return this
    }

    fun clientTrackingRange(pRange: Int): ETWrapper<T> {
        builder.clientTrackingRange(pRange)
        return this
    }

    fun updateInterval(interval: Int): ETWrapper<T> {
        builder.updateInterval(interval)
        return this
    }

    fun setUpdateInterval(interval: Int): ETWrapper<T> {
        builder.setUpdateInterval(interval)
        return this
    }

    fun setTrackingRange(range: Int): ETWrapper<T> {
        builder.setTrackingRange(range)
        return this
    }

    fun setShouldReceiveVelocityUpdates(value: Boolean): ETWrapper<T> {
        builder.setShouldReceiveVelocityUpdates(value)
        return this
    }

    fun setCustomClientFactory(customClientFactory: BiFunction<SpawnEntity?, World?, T>?): ETWrapper<T> {
        builder.setCustomClientFactory(customClientFactory)
        return this
    }
}